test/src/userobjects/ExampleShapeElementUserObject.C

```diff
@@ -0,0 +1,103 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "ExampleShapeElementUserObject.h"
+#include "libmesh/quadrature.h"
+
+template<>
+InputParameters validParams<ExampleShapeElementUserObject>()
+{
+  InputParameters params = validParams<ShapeElementUserObject>();
+  params.addRequiredCoupledVar("u", "first coupled variable");
+  params.addRequiredCoupledVar("v", "second coupled variable");
+  return params;
+}
+
+ExampleShapeElementUserObject::ExampleShapeElementUserObject(const InputParameters & parameters) :
+    ShapeElementUserObject(parameters),
+    _u_value(coupledValue("u")),
+    _u_var(coupled("u")),
+    _v_value(coupledValue("v")),
+    _v_var(coupled("v"))
+{
+}
+
+void
+ExampleShapeElementUserObject::initialize()
+{
+  _integral = 0.0;
+
+  // Jacobian term storage is up to the user. One option is using an std::vector
+  // We resize it to the total number of DOFs in the system and zero it out.
+  // WARNING: this can be large number (smart sparse storage could be a future improvement)
+  // if (_currently_computing_jacobian)
+    _jacobian_storage.assign(_subproblem.es().n_dofs(), 0.0);
```

Comment:
Hmmm... this is the big problem here.  That's not going to work past a few thousand dofs... especially in parallel (I see that you're doing a `gatherSum()` on this vector in `finalize()` below... that's going to kill you in parallel with a large problem).

Unfortunately... I don't really see another way either... because if you're coupling in an integral over the whole domain then this really is going to be a dense vector that is needed on every element in the whole domain.  Let me think about it a bit...